File: tests/client_compatibility/php/tests/QueryTest.php
<?php

namespace IrisPGWire\Tests;

use PHPUnit\Framework\TestCase;
use PDO;
use DateTime;

class QueryTest extends TestCase
{
    private PDO $pdo;

    protected function setUp(): void
    {
        $config = [
            'host' => getenv('PGWIRE_HOST') ?: 'localhost',
            'port' => getenv('PGWIRE_PORT') ?: '5432',
            'dbname' => getenv('PGWIRE_DATABASE') ?: 'USER',
            'user' => getenv('PGWIRE_USERNAME') ?: 'test_user',
            'password' => getenv('PGWIRE_PASSWORD') ?: 'test'
        ];

        $dsn = sprintf(
            'pgsql:host=%s;port=%s;dbname=%s',
            $config['host'],
            $config['port'],
            $config['dbname']
        );

        $this->pdo = new PDO(
            $dsn,
            $config['user'],
            $config['password'],
            [
                PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
                PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC
            ]
        );
    }

    public function testSelectConstant(): void
    {
        // GIVEN: Connected client
        // WHEN: Executing SELECT with constant
        $stmt = $this->pdo->query('SELECT 42 as answer');
        $result = $stmt->fetch();

        // THEN: Should return constant value
        $this->assertEquals(42, $result['answer']);
    }

    public function testSelectMultipleColumns(): void
    {
        // GIVEN: Connected client
        // WHEN: Selecting multiple columns
        $stmt = $this->pdo->query("SELECT 1 as num, 'hello' as text, 3.14 as pi");
        $result = $stmt->fetch();

        // THEN: All columns should be returned
        $this->assertEquals(1, $result['num']);
        $this->assertEquals('hello', $result['text']);
        $this->assertEqualsWithDelta(3.14, (float)$result['pi'], 0.001);
    }

    public function testSelectCurrentTimestamp(): void
    {
        // GIVEN: Connected client
        // WHEN: Selecting CURRENT_TIMESTAMP
        $stmt = $this->pdo->query('SELECT CURRENT_TIMESTAMP as ts');
        $result = $stmt->fetch();

        // THEN: Should return timestamp string
        $this->assertIsString($result['ts']);
        $this->assertNotEmpty($result['ts']);

        // Verify timestamp is reasonable (within last hour)
        $timestamp = strtotime($result['ts']);
        $now = time();
        $diff = abs($now - $timestamp);
        $this->assertLessThan(3600, $diff, 'Timestamp should be within last hour');
    }

    public function testSelectWithNull(): void
    {
        // GIVEN: Connected client
        // WHEN: Selecting NULL in comparison
        $stmt = $this->pdo->query('SELECT 1 WHERE 1=0');
        $result = $stmt->fetch();

        // THEN: Empty result set
        $this->assertFalse($result, 'Empty result set should return false');
    }

    public function testPreparedStatementSingleParam(): void
    {
        // GIVEN: Connected client
        // WHEN: Using prepared statement with parameter
        $stmt = $this->pdo->prepare('SELECT :value as result');
        $stmt->execute(['value' => 99]);
        $result = $stmt->fetch();

        // THEN: Value should be returned correctly
        $this->assertEquals(99, $result['result']);
    }

    public function testPreparedStatementMultipleParams(): void
    {
        // GIVEN: Connected client
        // WHEN: Using prepared statement with multiple parameters
        $stmt = $this->pdo->prepare('SELECT :num as num, :text as text');
        $stmt->execute([
            'num' => 42,
            'text' => 'test'
        ]);
        $result = $stmt->fetch();

        // THEN: All parameters should be returned correctly
        $this->assertEquals(42, $result['num']);
        $this->assertEquals('test', $result['text']);
    }

    public function testPreparedStatementWithNull(): void
    {
        // GIVEN: Connected client
        // WHEN: Testing NULL in comparison
        $stmt = $this->pdo->query('SELECT 1 WHERE NULL IS NULL');
        $result = $stmt->fetch();

        // THEN: Should return one row (NULL IS NULL is true)
        $this->assertEquals(1, $result['?column?']);
    }

    public function testStringWithSpecialCharacters(): void
    {
        // GIVEN: Connected client
        // WHEN: Querying string with special characters
        $testString = "hello'world\"with\\special";
        $stmt = $this->pdo->prepare('SELECT :text as result');
        $stmt->execute(['text' => $testString]);
        $result = $stmt->fetch();

        // THEN: Special characters should be preserved
        $this->assertEquals($testString, $result['result']);
    }

    public function testMultipleRowsResult(): void
    {
        // GIVEN: Connected client
        // WHEN: Querying multiple rows (UNION ALL)
        $stmt = $this->pdo->query("SELECT 1 as num, 'first' as text UNION ALL SELECT 2, 'second'");
        $results = $stmt->fetchAll();

        // THEN: Should return multiple rows
        $this->assertCount(2, $results);

        $this->assertEquals(1, $results[0]['num']);
        $this->assertEquals('first', $results[0]['text']);

        $this->assertEquals(2, $results[1]['num']);
        $this->assertEquals('second', $results[1]['text']);
    }

    public function testEmptyResultSet(): void
    {
        // GIVEN: Connected client
        // WHEN: Executing query with no results
        $stmt = $this->pdo->query('SELECT 1 WHERE 1=0');
        $results = $stmt->fetchAll();

        // THEN: Should return empty array
        $this->assertCount(0, $results);
    }

    public function testSequentialQueries(): void
    {
        // GIVEN: Connected client
        // WHEN: Executing multiple queries sequentially
        $stmt1 = $this->pdo->query('SELECT 1 as value');
        $result1 = $stmt1->fetch();

        $stmt2 = $this->pdo->query('SELECT 2 as value');
        $result2 = $stmt2->fetch();

        $stmt3 = $this->pdo->query('SELECT 3 as value');
        $result3 = $stmt3->fetch();

        // THEN: All queries should succeed
        $this->assertEquals(1, $result1['value']);
        $this->assertEquals(2, $result2['value']);
        $this->assertEquals(3, $result3['value']);
    }

    public function testBlobHandling(): void
    {
        // GIVEN: Connected client with binary data
        $binaryData = "\x00\x01\x02\x03\x04\x05";

        // WHEN: Storing and retrieving binary data
        $stmt = $this->pdo->prepare('SELECT :data as result');
        $stmt->bindValue('data', $binaryData, PDO::PARAM_LOB);
        $stmt->execute();
        $result = $stmt->fetch();

        // THEN: Binary data should be preserved
        $this->assertEquals($binaryData, $result['result']);
    }
}

File: tests/client_compatibility/ruby/test/query_test.rb
require 'minitest/autorun'
require 'pg'
require 'time'

class QueryTest < Minitest::Test
  def setup
    @conn = PG.connect(
      host: ENV['PGWIRE_HOST'] || 'localhost',
      port: ENV['PGWIRE_PORT'] || 5432,
      dbname: ENV['PGWIRE_DATABASE'] || 'USER',
      user: ENV['PGWIRE_USERNAME'] || 'test_user',
      password: ENV['PGWIRE_PASSWORD'] || 'test'
    )
  end

  def teardown
    @conn.close if @conn
  end

  def test_select_constant
    # GIVEN: Connected client
    # WHEN: Executing SELECT with constant
    result = @conn.exec('SELECT 42 as answer')

    # THEN: Should return constant value
    assert_equal 42, result[0]['answer'].to_i
  end

  def test_select_multiple_columns
    # GIVEN: Connected client
    # WHEN: Selecting multiple columns
    result = @conn.exec("SELECT 1 as num, 'hello' as text, 3.14 as pi")

    # THEN: All columns should be returned
    assert_equal 1, result[0]['num'].to_i
    assert_equal 'hello', result[0]['text']
    assert_in_delta 3.14, result[0]['pi'].to_f, 0.001
  end

  def test_select_current_timestamp
    # GIVEN: Connected client
    # WHEN: Selecting CURRENT_TIMESTAMP
    result = @conn.exec('SELECT CURRENT_TIMESTAMP as ts')

    # THEN: Should return timestamp string
    assert_kind_of String, result[0]['ts']
    refute_empty result[0]['ts']

    # Verify timestamp is reasonable (within last 24 hours)
    # Note: IRIS returns timestamps in UTC format without timezone indicator
    # Ruby's Time.parse treats it as local time, creating timezone offset
    # We use a more lenient 24-hour window to account for timezone differences
    timestamp = Time.parse(result[0]['ts'])
    now = Time.now
    diff = (now - timestamp).abs
    assert diff < 86400, 'Timestamp should be within last 24 hours'
  end

  def test_select_with_null
    # GIVEN: Connected client
    # WHEN: Selecting with WHERE clause that returns no rows
    result = @conn.exec('SELECT 1 WHERE 1=0')

    # THEN: Empty result set
    assert_equal 0, result.ntuples
  end

  def test_prepared_statement_single_param
    # GIVEN: Connected client
    # WHEN: Using prepared statement with parameter
    result = @conn.exec_params('SELECT $1::int as result', [99])

    # THEN: Value should be returned correctly
    assert_equal 99, result[0]['result'].to_i
  end

  def test_prepared_statement_multiple_params
    # GIVEN: Connected client
    # WHEN: Using prepared statement with multiple parameters
    result = @conn.exec_params('SELECT $1::int as num, $2::text as text', [42, 'test'])

    # THEN: All parameters should be returned correctly
    assert_equal 42, result[0]['num'].to_i
    assert_equal 'test', result[0]['text']
  end

  def test_prepared_statement_with_null
    # GIVEN: Connected client
    # WHEN: Testing NULL in comparison
    result = @conn.exec('SELECT 1 WHERE NULL IS NULL')

    # THEN: Should return one row (NULL IS NULL is true)
    assert_equal 1, result.ntuples
    assert_equal 1, result[0]['?column?'].to_i
  end

  def test_string_with_special_characters
    # GIVEN: Connected client
    # WHEN: Querying string with special characters
    test_string = "hello'world\"with\\special"
    result = @conn.exec_params('SELECT $1 as result', [test_string])

    # THEN: Special characters should be preserved
    assert_equal test_string, result[0]['result']
  end

  def test_multiple_rows_result
    # GIVEN: Connected client
    # WHEN: Querying multiple rows (UNION ALL)
    result = @conn.exec("SELECT 1 as num, 'first' as text UNION ALL SELECT 2, 'second'")

    # THEN: Should return multiple rows
    assert_equal 2, result.ntuples

    assert_equal 1, result[0]['num'].to_i
    assert_equal 'first', result[0]['text']

    assert_equal 2, result[1]['num'].to_i
    assert_equal 'second', result[1]['text']
  end

  def test_empty_result_set
    # GIVEN: Connected client
    # WHEN: Executing query with no results
    result = @conn.exec('SELECT 1 WHERE 1=0')

    # THEN: Should return empty result
    assert_equal 0, result.ntuples
  end

  def test_sequential_queries
    # GIVEN: Connected client
    # WHEN: Executing multiple queries sequentially
    result1 = @conn.exec('SELECT 1 as value')
    result2 = @conn.exec('SELECT 2 as value')
    result3 = @conn.exec('SELECT 3 as value')

    # THEN: All queries should succeed
    assert_equal 1, result1[0]['value'].to_i
    assert_equal 2, result2[0]['value'].to_i
    assert_equal 3, result3[0]['value'].to_i
  end

  def test_binary_data_handling
    # GIVEN: Connected client with binary data
    binary_data = "\x00\x01\x02\x03\x04\x05"

    # WHEN: Storing and retrieving binary data
    result = @conn.exec_params('SELECT $1 as result', [{ value: binary_data, format: 1 }])

    # THEN: Binary data should be preserved
    assert_equal binary_data, result[0]['result']
  end
end
